scripts/notebook-normalizer/qiskit_docs_notebook_normalizer/__init__.py

```diff
@@ -0,0 +1,221 @@
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+import argparse
+import nbformat
+import base64
+import shutil
+from typing import TypeGuard
+from subprocess import Popen, PIPE
+from pathlib import Path
+from dataclasses import dataclass
+from itertools import chain
+
+NOTEBOOK_PATHS = chain(
+    Path("docs").rglob("*.ipynb"),
+    Path("learning").rglob("*.ipynb"),
+)
+
+# Qiskit's QuantumCircuit.draw() results in Jupyter outputting both a `text/html` and
+# `text/plain` entry. The HTML entry has pre-applied formatting that makes sense in
+# a Jupyter notebook, but renders horribly in our app:
+# https://github.com/Qiskit/qiskit/blob/df379876ba10d6f490a96723b6dbbf723ec45d7a/qiskit/visualization/circuit/text.py#L761-L769
+#
+# So, we instead should render the `text/plain` entry rather than `text/html`.
+CIRCUIT_DRAW_HTML_PREFIX = '<pre style="word-wrap: normal;white-space: pre;background: #fff0;line-height: 1.1;font-family: &quot;Courier New&quot;,Courier,monospace">'
+
+# Types for extracted images
+
+
+@dataclass
+class SvgImage:
+    data: str
+    filepath: Path
+
+    def write(self):
+        self.filepath.write_text(self.data)
+
+
+@dataclass
+class RasterImage:
+    data: bytes
+    filepath: Path
+
+    def write(self):
+        self.filepath.write_bytes(self.data)
+
+
+Image = SvgImage | RasterImage
+
+
+# Result types for normalization process
+@dataclass
+class NormalizationNeeded:
+    changes = True
+    nb: nbformat.NotebookNode
+    images: list[Image]
+
+
+@dataclass
+class AlreadyNormalized:
+    changes = False
```

Comment:
More idiomatic is to make clear this is a classvar by using `ClassVar[bool]` here and line 63